### PR TITLE
Fix bug with middleware and router groups

### PR DESCRIPTION
**Pull request by orkosinha**
## Overview

There was a bug in the way `ApplicationAPI.ts` added different routers to Express. It turns out that adding a router with some middleware to a constant path (which we used to do for versioning) applies the same middleware to all routers added afterwards with the provided path. 

Because our path was constant for a given version in the `use` call (line 55 of ApplicationAPI.ts) it applied authentication middleware to routers we did not want to be authenticated, like the route to create users.

## Changes Made

This bug is present for a different reason in master. The master version worked like Pollo in that all routes by default had the authentication middleware. This is kind of the other way around where all routes by default do not have this middleware, and only routes that need to authenticate have the middleware implemented. 

Additionally, a new way to version routes has been added to `ApplicationAPI.ts`. This has a version function which needs to be overridden by API. This is to preserve versioning features. The repurposed `routerGroups()` is used to separate routes according to which middleware they use.

## Test Coverage

I tested using the pollo app for generating ID tokens with authentication and mitmproxy to intercept those ID tokens.

### CODE REVIEW

File: src/utils/Authenticate.ts [hunk 1]
```diff
@@ -27,13 +31,13 @@ async function ensureAuthenticated(
 ) {
   const bearerToken = parseToken(req, res);
 
-  if (bearerToken === '') return next(true);
```
Comment (orkosinha) [inline src/utils/Authenticate.ts:30]: I forgot to remove these guys in the last PR. They were for debugging and print true when authentication fails.

**Review by seancorc — COMMENTED** [review]
Is the intended use that all the routers independently get to manage which middleware they are using? Because when I set the middleware of the GetClubsRouter to `[]` I still receive an `Authorization header missing` which seemingly means the auth middleware is still being used.

**Comment by orkosinha** [comment]
> Is the intended use that all the routers independently get to manage which middleware they are using? Because when I set the middleware of the GetClubsRouter to `[]` I still receive an `Authorization header missing` which seemingly means the auth middleware is still being used.

Router groups are what control which middleware is used. However, if the middleware isn't specified in any of the `/user/` routes it should be added. We could potentially make this more dynamic with each router/route using it's own middleware but I don't think that's necessary.

I might just make a class for authenticated routes extending the regular ApplicationRouter because it's kind of annoying to specify middleware every time.

**Review by seancorc — APPROVED** [review]
Good stuff 🔥 , your an AppDev Utils god

**Review by williambaibai — APPROVED** [review]
LGTM!

**Pull request merged**